src/hookintrprtr.h

```diff
@@ -0,0 +1,183 @@
+** interpreter every time statements are executed. Note that the existence
+** of any hooks slows GAP down measurably, so don't leave them in if you
+** don't need them, and try to make it clear to users they are activated,
+** and how to deactivate them.
+**
+** There are four functions:
+**
+** * 'visitStat' is called for every visited Stat (and Expr) from the
+**    GAP bytecode.
+** * 'enterFunction' and 'leaveFunction' are called whenever a function
+**    is entered, or left. This is passed the function which is being
+**    entered (or left)
+** * 'registerStat' is called whenever a statement is read from a text
+**    file. Note that you will only see files which are read while your
+**    hooks are running.
+** * 'hookName' is a string is used in debugging messages to describe
+**    the currently active hooks.
+**
+** This is a sharp tool -- use with care! Look at 'profiling.c', and
+** the 'debugger' package for guidance on usage, in particular look
+** at FILENAMEID_STAT, FILENAME_STAT and LINE_STAT to find out which
+** statement is running.
+**
+** Remember if you run GAP code during any of these functions, it will
+** reinvoke your hooks!
+*/
+
+struct InterpreterHooks {
+    void (*visitStat)(Stat stat);
+    void (*enterFunction)(Obj func);
+    void (*leaveFunction)(Obj func);
+    void (*registerStat)(Stat stat);
+    const char * hookName;
+};
+
+
+enum { HookCount = 6 };
+
+extern struct InterpreterHooks * activeHooks[HookCount];
+
+Int ActivateHooks(struct InterpreterHooks * hook);
+Int DeactivateHooks(struct InterpreterHooks * hook);
+
+/****************************************************************************
+**
+** A struct to represent the hooks allowed into printing functions
+**
+**
+** This struct represents a list of functions which will be called whenever
+** a statement or expression is printed. They can be used to provide
+** simple customisation of printing. At the moment they are used by
+** profiling.c, to mark statements which have been executed.
+** Look at that code to get an idea how to use them.
+*/
+struct PrintHooks {
+    void (*printStatPassthrough)(Stat stat);
+    void (*printExprPassthrough)(Expr stat);
+    char * hookName;
+};
+
+void ActivatePrintHooks(struct PrintHooks * hook);
+void DeactivatePrintHooks(struct PrintHooks * hook);
+
+/****************************************************************************
+**
+** We can only store a fixed number of files, and files, in statements.
+**
+** Whenever an overflow occurs we do not mark that statement with it's line
+** or file, and call these functions to record the fact information is
+** missing, so we can inform users later.
+*/
+
+void ReportLineNumberOverflowOccured();
+void ReportFileNumberOverflowOccured();
+
+/****************************************************************************
+**
+** We need this to be in the header, so they can be inlined away. The only
```

Comment:
Reading this, it is unclear to me what "this" refers to. Perhaps "the following declarations"? Or how about: "The following things are for internal use only. The only reason they appear in this header is because they are used by multiple files."